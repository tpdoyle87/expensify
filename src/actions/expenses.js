import database  from '../firebase/firebase';

// component calls action generator
// action generator returns object
// componnent dispatches object
// return store changes

// component calls action generator
// action generator returns a function
// componnent dispatches function (?)
// function runs ( has the ability to dispatch other actions and do whatever it wants)

// Add_expense
export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = {description, note, amount, createdAt}
    database.ref(`users/${uid}/expenses`).push(expense).then((snapshot) => {
      dispatch(addExpense({
        id: snapshot.key,
        ...expense
      }));
    })
  }
}
// Remove_expense
export const removeExpense = ({id}) => ({
  type: "REMOVE_EXPENSE",
  id: id
})

export const startRemoveExpense = ({ id }) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({id}))
    })
  }
}
// Edit_expense
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
})

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates))
    })
  }
}

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
      const expenses = [];
      snapshot.forEach((expense) => {
        expenses.push({
          id: expense.key,
          ...expense.val()
        })
      })
      dispatch(setExpenses(expenses))
    })
  }
}
