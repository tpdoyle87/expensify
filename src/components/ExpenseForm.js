import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import uuid from 'uuid'

const date = moment();
export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : 0,
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calanderFocused: false,
      error: undefined
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value
    this.setState(() => ({ description }))
  }

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }))
  }
  onAmountChange = (e) => {
    const amount = e.target.value
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  }
  onDateChange = (createdAt) => {
    if (createdAt) {
    this.setState(() => ({ createdAt }))
  }}
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calanderFocused: focused }))
  }
  onSubmit = (e) => {
    e.preventDefault()
    if (!this.state.description || !this.state.amount ) {
      //set Error State to "Please provide description and amount"
      this.setState(() => ({
        error: "Please Provide a description and amount"
      }))
    } else {
      // clear the error
      this.setState(() => ({
        error: undefined
      }))
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      })
    }
  }

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        {this.state.description === '' ? <h3>Add Expense</h3> : <h3>Edit Expense</h3>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="description"
            value={this.state.description}
            onChange={this.onDescriptionChange}
            autoFocus
          />
          <input
            type="text"
            placeholder="amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calanderFocused}
            onFocusChange={this.onFocusChange}
            id={uuid()}
            numberOfMonths={1}
            isOutsideRange={() => false}
            hideKeyboardShortcutsPanel={true}
          />
          <textarea
            placeholder="add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          >

          </textarea>
          <button>{this.state.description === '' ? "Add Expense" : "Edit Expense"}</button>
        </form>
      </div>
    )
  }
}

