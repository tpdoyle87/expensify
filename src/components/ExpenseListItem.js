import React from 'react'

const ExpenseListItem = ({description, amount, createdAt}) => (
  <div>
      <h3>Description: {description}</h3>
      <p>{amount}-{createdAt}</p>
  </div>
)

export default ExpenseListItem
