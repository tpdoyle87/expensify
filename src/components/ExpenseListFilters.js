import React from 'react';
import uuid from 'uuid'
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters'
import { DateRangePicker } from "react-dates"
import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css';


class ExpenseListFilters extends React.Component {
  state = {
    calenderFocused: null
  }

  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate))
    this.props.dispatch(setEndDate(endDate))
  }
  render() {
    return(
  <div>
    <input
      type="text"
      value={this.props.filters.text}
      onChange={(e) => {
      this.props.dispatch(setTextFilter(e.target.value))
    }}
    />
    <select
      value={this.props.filters.sortBy}
      onChange={(e) => {
        if (e.target.value === "amount") {
          this.props.dispatch(sortByAmount(e.target.value))
        } else {
          this.props.dispatch(sortByDate(e.target.value))
        }
    }}>
      <option value="date">Date</option>
      <option value="amount">Amount</option>
    </select>
    <DateRangePicker
      startDate={this.props.filters.startDate}
      endDate={this.props.filters.endDate}
      startDateId={uuid()}
      endDateId={uuid()}
      onDatesChange={ this.onDatesChange}
      focusedInput={this.state.calenderFocused}
      onFocusChange={calenderFocused => this.setState({calenderFocused})}
      numberOfMonths={1}
      isOutsideRange={() => false}
      showClearDates={true}
    />
  </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  }
}



export default connect(mapStateToProps)(ExpenseListFilters)
