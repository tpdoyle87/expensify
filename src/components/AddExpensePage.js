import React from 'react'
import { connect } from 'react-redux'

import ExpenseForm from './ExpenseForm'
import { startAddExpense } from '../actions/expenses'

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
      this.props.startAddExpense(expense)
      this.props.history.push("/")
  }

  render() {
    return (
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
