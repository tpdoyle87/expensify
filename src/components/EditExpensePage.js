import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm'
import { startEditExpense } from '../actions/expenses'
import { startRemoveExpense } from '../actions/expenses'

class EditExpensePage extends React.Component {
  onEdit = (expense) => {
    this.props.dispatch(startEditExpense(this.props.expense.id, expense))
    this.props.history.push("/dashboard")
  }

  onRemove = () => {
    this.props.dispatch(startRemoveExpense({ id: this.props.expense.id }))
    this.props.history.push('/dashboard')
  }
  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onEdit}
        />
        <button onClick={this.onRemove}
        >
          Remove
      </button>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
}



export default connect(mapStateToProps)(EditExpensePage);
