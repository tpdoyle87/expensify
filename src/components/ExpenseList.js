import React from 'react'
import { connect } from 'react-redux'

import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'
import { setExpenses } from '../actions/expenses'

class ExpenseList extends React.Component {

  render() {
    return (
      <div>
        {
          this.props.expenses.length > 0 ? (
            this.props.expenses.map((expense) => {
              return <ExpenseListItem key={expense.id} {...expense} />;
            })
          ) : (
              <p>No expenses</p>
            )
        }
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
    filters: state.filters
  }
};


export default connect(mapStateToProps)(ExpenseList);

