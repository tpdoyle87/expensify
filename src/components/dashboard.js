import React from 'react'
import ConnectedExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ConnectedExpenseSummary from './ExpenseSummary'

export default class Dashboard extends React.Component {
  render() {
    return(
      <div>
        <ConnectedExpenseSummary />
        <ExpenseListFilters />
        <ConnectedExpenseList />
      </div>
    );
  }
}
