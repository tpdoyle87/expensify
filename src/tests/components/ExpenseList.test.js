import { ExpenseList } from '../../components/ExpenseList'
import React from 'react'
import { shallow } from 'enzyme'
import expenses from '../fixtures/expenses'

test('it should render expense list with expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses}/>)
  expect(wrapper).toMatchSnapshot();
})

test('it should render expense list with empty message', () => {
  const wrapper = shallow(<ExpenseList />)
  expect(wrapper).toMatchSnapshot();
})
