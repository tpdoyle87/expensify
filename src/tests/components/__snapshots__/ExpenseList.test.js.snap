// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render expense list with empty message 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`it should render expense list with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={3000}
    createdAt={1}
    description="water"
    id={1}
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={9000}
    createdAt={2}
    description="rent"
    id={2}
    key="2"
    note="bye"
  />
  <ExpenseListItem
    amount={5000}
    createdAt={3}
    description="gas"
    id={3}
    key="3"
    note="hi"
  />
</div>
`;
